src/profile-logic/line-timings.js

```diff
@@ -0,0 +1,263 @@
+ * For each stack in the stack table and each line in the file, we answer the
+ * question "Does this stack contribute to line X's self time? Does it contribute
+ * to line X's total time?"
+ * Each stack can only contribute to one line's self time: the line of the stack's
+ * own frame.
+ * But each stack can contribute to the total time of multiple lines: All the lines
+ * in the file that are encountered by any of the stack's ancestor stacks.
+ * E.g if functions A, B and C are all in the same file, then a stack with the call
+ * path [A, B, C] will contribute to the total time of 3 lines:
+ *   1. The line in function A which has the call to B,
+ *   2. The line in function B which has the call to C, and
+ *   3. The line in function C that is being executed at that stack (stack.frame.line).
+ *
+ * This last line is the stack's "self line".
+ * If there is recursion, and the same line is present in multiple frames in the
+ * same stack, the line is only counted once - the lines are stored in a set.
+ *
+ * The returned StackLineInfo is computed as follows:
+ *   selfLine[stack]:
+ *     For stacks whose stack.frame.func.file is the given file, this is stack.frame.line.
+ *     For all other stacks this is null.
+ *   stackLines[stack]:
+ *     For stacks whose stack.frame.func.file is the given file, this is the stackLines
+ *     of its prefix stack, plus stack.frame.line added to the set.
+ *     For all other stacks this is the same as the stackLines set of the stack's prefix.
+ */
+export function getStackLineInfoNonInverted(
+  stackTable: StackTable,
+  frameTable: FrameTable,
+  funcTable: FuncTable,
+  fileNameStringIndex: IndexIntoStringTable
+): StackLineInfo {
+  // "self line" == "the line which a stack's self time is contributed to"
+  const selfLineForAllStacks = [];
+  // "total lines" == "the set of lines whose total time this stack contributes to"
+  const totalLinesForAllStacks = [];
+
+  // This loop takes advantage of the fact that the stack table is topologically ordered:
+  // Prefix stacks are always visited before their descendants.
+  // Each stack inherits the "total" lines from its parent stack, and then adds its
+  // self line to that set. If the stack doesn't have a self line in the file, we just
+  // re-use the prefix's set object without copying it.
+  for (let stackIndex = 0; stackIndex < stackTable.length; stackIndex++) {
+    const frame = stackTable.frame[stackIndex];
+    const prefixStack = stackTable.prefix[stackIndex];
+    const func = frameTable.func[frame];
+    const fileNameStringIndexOfThisStack = funcTable.fileName[func];
+
+    let selfLine: LineNumber | null = null;
+    let totalLines: Set<LineNumber> | null =
+      prefixStack !== null ? totalLinesForAllStacks[prefixStack] : null;
+
+    if (fileNameStringIndexOfThisStack === fileNameStringIndex) {
+      selfLine = frameTable.line[frame];
+      if (selfLine !== null) {
+        if (totalLines === null) {
```

Comment:
Doesn't have to be the root - it could also be that all ancestors were in different files, and this is just the first hit for the requested file. I'll add a comment.